### PR TITLE
Feature/rolling shutter smart factors

### PR DESCRIPTION
**Pull request by lucacarlone**
- [x] added jacobians for interpolate (@shteren1)
- [x] added projection factor for rolling shutter camera (@shteren1)
- [x] added tests for projection factor
- [x] added Smart projection factor with rolling shutter
- [x] added tests for Smart projection factor with rolling shutter

### CODE REVIEW

**Comment by lucacarlone** [comment]
@shteren1 : I added the tests and applied minor changes to your implementation (which already looked great and worked out of the box!). The ProjectionFactor part is pretty much ready. Now I will move to developing the smart factor version. Let me know if you have comments.

File: gtsam/base/Lie.h
Comment (shteren1) [inline gtsam/base/Lie.h:20]: Yotam Stern is my name :)

Comment (lucacarlone) [reply gtsam/base/Lie.h:20]: no worries! I already fixed that locally - it will be part of the next commit!

**Comment by shteren1** [comment]
Thanks Luca! I know the factor works, I've used it extensively to create stunning vision only rolling shutter SLAM's with loop closures and all.

I did compile on the original version I developed it on - 4.0.3 stable, not sure why it failed to compile on the PR, thanks for taking care of it. 

@dellaert @lucacarlone I also have a version of this factor that accepts body-to-sensor pose3 matrix and optimizes it, incase of a multi camera setup with unknown calibration between cameras. I thought it was very nitche'y so I didn't add it, but if you see value it in I can add it also.

**Comment by lucacarlone** [comment]
> Thanks Luca! I know the factor works, I've used it extensively to create stunning vision only rolling shutter SLAM's with loop closures and all.
> 
> It did compile on the original version I developed it on - 4.0.3 stable, not sure why it failed to compile on the PR, thanks for taking care of it.
> 
> @dellaert @lucacarlone I also have a version of this factor that accepts body-to-sensor pose3 matrix and optimizes it, incase of a multi camera setup with unknown calibration between cameras. I thought it was very nitche'y so I didn't add it, but if you see value it in I can add it also.

cool! out of curiosity: what tool did you use to calibrate the row readout time? also: did you still use a RANSAC-like preprocessing to filter out outliers before the optimization?

I will leave it to @dellaert to decide if the other body-to-sensor factor is of interest. I'll be happy to help with the tests if needed.

**Comment by shteren1** [comment]
I work with a specific camera for which I know all the intrinsic parameters and read times so I didn't require any calibration for that.

As far as the SLAM goes, I used a simple matrix estimation between consecutive frames to initialize the SLAM, the essential matrix part used the RANSAC + essential matrix calculation from opencv. I know its possible to build a similar optimizer with gtsam but opencv's one has a complete simple usage solution that works fast. I also used SIFT extraction and matching from opencv to build the SLAM graph.

**Comment by lucacarlone** [comment]
> I work with a specific camera for which I know all the intrinsic parameters and read times so I didn't require any calibration for that.
> 
> As far as the SLAM goes, I used a simple essential matrix estimation between consecutive frames to initialize the SLAM, the essential matrix part used the RANSAC + essential matrix calculation from opencv. I know its possible to build a similar optimizer with gtsam but opencv's one has a complete simple usage solution that works fast. I also used SIFT extraction and matching from opencv to build the SLAM graph.

great - thanks for sharing!

**Comment by shteren1** [comment]
@lucacarlone So I had a look at what those smart factors are all about and read the article, that is fantastic work!

And adding support for smart rolling shutter factor is just superb, It's going to enable real time SLAM on many light and cheap platforms!

**Comment by lucacarlone** [comment]
> @lucacarlone So I had a look at what those smart factors are all about and read the article, that is fantastic work!
> 
> And adding support for smart rolling shutter factor is just superb, It's going to enable real time SLAM on many light and cheap platforms!

thanks, @shteren1 !!  I agree it would be cool to see them running on hw - in the past, we have seen that the Schur complement trick gives us up to 10x speedup compared to projection factors that also optimize for the 3D points. Let's see if the implementation in this PR is as efficient!

**Comment by lucacarlone** [comment]
@dellaert @shteren1 : this is good to go. Let me know if you have comments!

**Comment by shteren1** [comment]
@lucacarlone it looks good I went over the code, very impressive work with the jacobians there. looking forward to using it once its merged.

File: gtsam/base/Lie.h
Comment (dellaert) [inline gtsam/base/Lie.h:333]: How about giving these informative names e.g. `XinvY_H_x` and do a simple multiplication in lines 346 and 347. It might also be interesting to look at the generic implementations of compose and between:
```
  Class compose(const Class& g, ChartJacobian H1,
      ChartJacobian H2 = boost::none) const {
    if (H1) *H1 = g.inverse().AdjointMap();
    if (H2) *H2 = Eigen::Matrix<double, N, N>::Identity();
    return derived() * g;
  }

  Class between(const Class& g, ChartJacobian H1,
      ChartJacobian H2 = boost::none) const {
    Class result = derived().inverse() * g;
    if (H1) *H1 = - result.inverse().AdjointMap();
    if (H2) *H2 = Eigen::Matrix<double, N, N>::Identity();
    return result;
  }
```
from which you can see the `H2` in both cases is identity, allowing some simplification below as well.

File: gtsam/geometry/CameraSet.h
Comment (dellaert) [inline gtsam/geometry/CameraSet.h:201]: differently

File: gtsam/geometry/CameraSet.h
Comment (dellaert) [inline gtsam/geometry/CameraSet.h:285]: Why? `SchurComplementAndRearrangeBlocks<3,12,6>` reads better to me than `SchurComplementAndRearrangeBlocks_3_12_6`, and if you used fixed-size (or mixed fixed-dynamic) matrices the dimensions might be derivable as well...

File: gtsam/geometry/CameraSet.h
Comment (dellaert) [inline gtsam/geometry/CameraSet.h:202]: I think I need more explanation here

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/ProjectionFactorRollingShutter.h:39]: Since not templated, can we move most implementation to cpp file?

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/ProjectionFactorRollingShutter.h:29]: Did you format all new code with Google style? Suspiciously long lines here.

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/ProjectionFactorRollingShutter.h:34]: Not loving `interp_param`. Can we just call this `alpha`? Also note in docs that alpha is between 0.0 and 1.0.

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:95]: same. alpha?

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (dellaert) [inline gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:13]: Can we add a .cpp file to make this file easier to read?

**Review by dellaert — CHANGES_REQUESTED** [review]
Very cool. Would like to take another look so marking `Request Changes`, but I think it's close.

**Comment by lucacarlone** [comment]
@dellaert @shteren1 : thanks for the comments. I'll implement them in the next few days and ping you when ready.

**Comment by shteren1** [comment]
@dellaert regarding alpha greater than 1, it is required for the edge case of the last pose in the graph.
since you don't have the next pose to interpolate with, you must do a small extrapolation using the previous pose.
I tried using a generic projection factor just for the last pose in the graph and It's position got slightly distorted. I think using a small extrapolation is a better solution.

**Comment by lucacarlone** [comment]
> @dellaert regarding alpha greater than 1, it is required for the edge case of the last pose in the graph.
> since you don't have the next pose to interpolate with, you must do a small extrapolation using the previous pose.
> I tried using a generic projection factor just for the last pose in the graph and It's position got slightly distorted. I think using a small extrapolation is a better solution.

I agree with @shteren1 that having the option to extrapolate would be useful. I plan to use this in Kimera, and if we can extrapolate the integration in our code will be straightforward.

**Comment by dellaert** [comment]
Yeah, no worries: just document: "typically \in [0,1] but can also be used to extrapolate before pose A or after pose B." ?

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (lucacarlone) [reply gtsam_unstable/slam/ProjectionFactorRollingShutter.h]: done - let me know if you prefer I move more code to cpp

File: gtsam/base/Lie.h
Comment (lucacarlone) [reply gtsam/base/Lie.h]: done

File: gtsam/base/Lie.h
Comment (lucacarlone) [reply gtsam/base/Lie.h]: great suggestion - done

File: gtsam/geometry/CameraSet.h
Comment (lucacarlone) [reply gtsam/geometry/CameraSet.h]: done

File: gtsam/geometry/CameraSet.h
Comment (lucacarlone) [reply gtsam/geometry/CameraSet.h:286]: do you have an example to clarify how to derive the dimension?

Regarding "SchurComplementAndRearrangeBlocks<3,12,6>": it does not compile on my side and I get "error: 
reference to overloaded function could not be resolved; did you mean to call it?". Unfortunately, I was not able to get around that error.

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (lucacarlone) [reply gtsam_unstable/slam/ProjectionFactorRollingShutter.h:29]: I did - comments still look quite long though

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (lucacarlone) [reply gtsam_unstable/slam/ProjectionFactorRollingShutter.h]: done

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (lucacarlone) [reply gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h]: done

File: gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h
Comment (lucacarlone) [reply gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:13]: this is templated. What do you suggest moving?

Comment (dellaert) [reply gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:13]: Ah, never mind.

Comment (lucacarlone) [reply gtsam_unstable/slam/SmartProjectionPoseFactorRollingShutter.h:13]: happy to template the basic (non-smart factor) if you like, to keep symmetry.

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (dellaert) [reply gtsam_unstable/slam/ProjectionFactorRollingShutter.h:29]: I reformatted with Google style and lots changed. Check your IDE :-)

Comment (lucacarlone) [reply gtsam_unstable/slam/ProjectionFactorRollingShutter.h:29]: Screen Shot 2021-08-28 at 5 37 38 PM

what do you use on your side?

**Review by dellaert — APPROVED** [review]
I fixed the template issue and formatted new files with clang-format, Google style. Will merge after CI is complete.

**Comment by dellaert** [comment]
@shteren1 could you add unit tests for Lie::interpolate though? It's a fairly complex Jacobian calculation which was added but has no direct unit test. See #864

File: gtsam_unstable/slam/ProjectionFactorRollingShutter.h
Comment (dellaert) [reply gtsam_unstable/slam/ProjectionFactorRollingShutter.h]: clang-format, Google style.

**Comment by lucacarlone** [comment]
> I fixed the template issue and formatted new files with clang-format, Google style. Will merge after CI is complete.

great - thanks!

**Comment by lucacarlone** [comment]
> @shteren1 could you add unit tests for Lie::interpolate though? It's a fairly complex Jacobian calculation which was added but has no direct unit test. See #864

I think I added unit tests for that in testPose3. should I move to testLie?

**Comment by dellaert** [comment]
PS @lucacarlone @shteren1 part of the issue was some arguments were not const&, just const. Somehow that prevented template instantiation. But changing to const& also helps with performance, as only a pointer is passed.

**Comment by lucacarlone** [comment]
> PS @lucacarlone @shteren1 part of the issue was some arguments were not const&, just const. Somehow that prevented template instantiation. But changing to const& also helps with performance, as only a pointer is passed.

thanks for figuring that out! (btw: I also figured out the clang formatting issue - thanks for mentioning the style was off)

**Comment by dellaert** [comment]
Ok, @lucacarlone . Merged, so you can now merge develop into your other PR branches. Hopefully not too many conflicts due to formatting.

**Comment by lucacarlone** [comment]
the merge was not too bad so far - thanks @dellaert !

**Pull request merged**
